Revamp WebCodecs documentation (#43697)
mdn/content: files/en-us/web/api/webcodecs_api/index.md

## Concepts

The WebCodecs API provides browser-native interfaces to represent raw video frames, encoded video frames, as well as raw and encoded audio.

|             | Video                            | Audio                            |
| ----------- | -------------------------------- | -------------------------------- |
| **Raw**     | {{domxref("VideoFrame")}}        | {{domxref("AudioData")}}         |
| **Encoded** | {{domxref("EncodedVideoChunk")}} | {{domxref("EncodedAudioChunk")}} |

The WebCodecs API also introduces the {{domxref("VideoDecoder")}} and {{domxref("VideoEncoder")}} interfaces, which transform `EncodedVideoChunk` objects into `VideoFrame` objects and vice-versa.

![VideoEncoder and VideoDecoder](video-encoder-decoder.png)

Likewise, the WebCodecs API also introduces the {{domxref("AudioDecoder")}} and {{domxref("AudioEncoder")}} interfaces, which transform `EncodedAudioChunk` objects into `AudioData` objects and vice-versa.

![AudioEncoder and AudioDecoder](audio-encoder-decoder.png)

> There is a 1:1 correspondence between the raw and encoded versions of each media type. Decoding n `EncodedVideoChunk` objects will yield exactly n `VideoFrame` objects (same with audio).

Reviewer: I think exceptions to this exist, but maybe not worth detailing here.

### Codecs

A codec is a specific algorithm for encoding (compressing) and decoding (decompressing) video and audio. There are several industry standard codecs for video, and a separate set of codecs for audio. Here are the major ones supported by the WebCodecs API:

#### Video codecs

- H.264 (AVC)
  - : The most widely supported video codec. Most MP4 files use H.264.
- VP9
  - : Open source, developed by Google. Better compression than H.264. Commonly used on YouTube and in WebM files.
- AV1
  - : The newest open source codec, with better compression than VP9. Broad decoder support; hardware encoder support is still limited.
- H.265 (HEVC)
  - : Better compression than H.264, but with significant gaps in browser support outside of Apple platforms.

#### Audio codecs

- Opus
  - : Open source, low-latency. The recommended choice for most WebCodecs audio encoding.
- AAC
  - : Widely supported. Common in MP4 files.
- MP3
  - : Broadly supported for decoding, but not available as an encoder in WebCodecs.
- PCM
  - : Uncompressed audio. No quality loss, but large file sizes.

> The WebCodecs specification supports a particular set of codecs, and individual devices and browsers may only support a subset of those. Encoders and decoders must be configured with fully specified codec strings (such as `"vp09.00.40.08.00"` for VP9 or `"avc1.4d0034"` for H.264) instead of ambiguous codec names like `"vp9"` or `"h264"`. For a full list of codec strings and their browser support, see the [Codec Support Table](https://webcodecsfundamentals.org/datasets/codec-support-table/) (webcodecsfundamentals.org), or the [Codec selection guide](/en-US/docs/Web/API/WebCodecs_API/Codec_selection) for guidance on choosing the right codec string.

Reviewer 2: Lets direct to the guidance first.

```suggestion
The WebCodecs specification supports a particular set of codecs, and individual devices and browsers may only support a subset of those. Encoders and decoders must be configured with fully specified codec strings (such as `"vp09.00.40.08.00"` for VP9 or `"avc1.4d0034"` for H.264) instead of ambiguous codec names like `"vp9"` or `"h264"`. The [Codec selection guide](/en-US/docs/Web/API/WebCodecs_API/Codec_selection) provides guidance on choosing an appropriate codec string (see the [Codec Support Table](https://webcodecsfundamentals.org/datasets/codec-support-table/) (webcodecsfundamentals.org) for a complete list of codec strings and their browser support). 
```

### Muxing and demuxing

The WebCodecs API only deals with encoding and decoding, with encoded chunks just representing binary data. It does not provide a built-in way to read `EncodedVideoChunk` objects from a video file, or write them to a playable video file.

Reading encoded chunks from a video file is a completely different process called demuxing, and to fetch `EncodedVideoChunk` objects from a video file, you will need to use a demuxing library such as [Mediabunny](https://mediabunny.dev/) or [web-demuxer](https://github.com/bilibili/web-demuxer).

![Demuxer](decoder-demuxer.png)

These libraries will follow the video container specifications (e.g., webm, mp4) to extract the track data and byte offsets for each encoded chunk, and provide methods for extracting the actual chunks from the raw file.

> Likewise, to write to a playable video file, you will need a muxing library, with [Mediabunny](https://mediabunny.dev/) being the primary option for muxing. Muxing libraries will handle formatting the binary encoded data, and placing it in the correct byte position in the output file stream according to the container specification, so that the output video is playable.

Reviewer 2: ```suggestion
Likewise, to write to a playable video file, you will need a muxing library, with [Mediabunny](https://mediabunny.dev/) being the primary option. Muxing libraries handle formatting the binary encoded data, and placing it in the correct byte position in the output file stream according to the container specification, so that the output video is playable.
```

You can find more information on muxing and demuxing in the [Muxing and Demuxing guide](https://webcodecsfundamentals.org/basics/muxing/) (webcodecsfundamentals.org).

## Guides

- [Video processing concepts](/en-US/docs/Web/API/WebCodecs_API/Video_processing_concepts)

>   - : A brief primer on video processing, covering codecs and containers, muxing and demuxing, and provides conceptual information to understand how the WebCodecs API implements these concepts.

Reviewer 2: ```suggestion
  - : A brief primer on video processing, covering codecs and containers, muxing and demuxing, and conceptual information that explains how the WebCodecs API implements these concepts.
```

{{DefaultAPISidebar("WebCodecs API")}}{{AvailableInWorkers("window_and_dedicated")}}

The **WebCodecs API** enables web developers to encode and decode video and audio in the browser efficiently (using hardware acceleration) and with very low-level control (processing on a per-frame basis).

> It is useful for web applications that do heavy media processing, or which require low-level control over the way media is encoded, such as browser-based video and audio editing, as well as live-streaming and video conferencing.

Reviewer 2: I like short choppy sentences. 
```suggestion
It is useful for web applications that do heavy media processing, or which require low-level control over the way media is encoded.
This includes browser-based video and audio editing, as well as live-streaming and video conferencing.
```

## Why WebCodecs exists

There are other APIs which use media codecs internally, such as the [MediaRecorder API](/en-US/docs/Web/API/MediaRecorder) and the [WebRTC API](/en-US/docs/Web/API/WebRTC_API), but these lack the low-level (per-frame) control required by some applications.

Previously, developers used WebAssembly ports of ffmpeg, but these lack true hardware acceleration capabilities, and are difficult to integrate with other key APIs like the File API for working with large files efficiently.

> WebCodecs was designed to enable low-level, hardware-accelerated media processing, for applications such as high-performance streaming and video editing, which were not well served by the existing APIs.

Reviewer 2: I wonder if this information is actually still useful/relevant? 

Specifically we're past the point where people are wondering if it is worth having, and it repeats a lot of the previous information. Maybe discard?

## Concepts

The WebCodecs API provides browser-native interfaces to represent raw video frames, encoded video frames, as well as raw and encoded audio.

|             | Video                            | Audio                            |
| ----------- | -------------------------------- | -------------------------------- |
| **Raw**     | {{domxref("VideoFrame")}}        | {{domxref("AudioData")}}         |
| **Encoded** | {{domxref("EncodedVideoChunk")}} | {{domxref("EncodedAudioChunk")}} |

The WebCodecs API also introduces the {{domxref("VideoDecoder")}} and {{domxref("VideoEncoder")}} interfaces, which transform `EncodedVideoChunk` objects into `VideoFrame` objects and vice-versa.

![VideoEncoder and VideoDecoder](video-encoder-decoder.png)

Likewise, the WebCodecs API also introduces the {{domxref("AudioDecoder")}} and {{domxref("AudioEncoder")}} interfaces, which transform `EncodedAudioChunk` objects into `AudioData` objects and vice-versa.

![AudioEncoder and AudioDecoder](audio-encoder-decoder.png)

> Generally there is a 1:1 correspondence between the raw and encoded versions of each media type. Decoding n `EncodedVideoChunk` objects will yield n `VideoFrame` objects (same with audio).

Reviewer 2: Intersting.

```suggestion
Generally there is a 1:1 correspondence between the raw and encoded versions of each media type. Decoding a number of `EncodedVideoChunk` objects will yield the same number of `VideoFrame` objects (and this is also true for audio).
```

### Video

A `VideoFrame` represents a video frame, and is tied both to actual pixel data on the device's graphics memory, as well as metadata such as the timestamp and duration (in microseconds), format and resolution. A `VideoFrame` can be constructed from any image source, and can also be rendered to a {{domxref("Canvas")}} using any of the canvas rendering methods.

> `EncodedVideoChunk` represents the encoded version of the same frame, tied to binary data in regular memory and the same metadata, with one key additional field: `type`, which can be "key" or "delta", representing whether it corresponds to a [key frame](https://webcodecsfundamentals.org/basics/encoded-video-chunk/#key-frames).

Reviewer 2: ```suggestion
`EncodedVideoChunk` represents the encoded version of the same frame, tied to binary data in regular memory and the same metadata.
The only difference is that it has one additional field: `type`, which can be "key" or "delta", representing whether  or not it corresponds to a [key frame](https://webcodecsfundamentals.org/basics/encoded-video-chunk/#key-frames).
```

> ![VideoFrame and EncodedVideoChunk](video-frame.png)

Reviewer 2: What am I supposed to be taking away from the numbers 900kb and 9kb on the images - encoded version is compressed by a factor of 100?

Reviewer 2: Or to put it another way, I bet you are. The key frame is probably 900Kb, but subsequent delta frames are 9K. We should make some kind of comment that captures this may point - the chunk data is compressed.

That is after all the point of why we have it, and I don't think we explained that up front - why do we "need" the two object types.

Author: The original intent was the former, though actually now that you mention it the diagram conflates 2 things, if a raw frame is 900kb, a typical key frame might be 10x smaller and the delta frames might be ~100x smaller. I updated the image to show 90kb for the encoded version and added some additional text below to clarify that the encoded chunk data is the compressed version of the raw frame.

### Audio

> An `AudioData` object represents a number of individual audio samples (1024 is a typical number). Audio sample data can be extracted as a {{jsxref("Float32Array")}} via the `copyTo` method. There is no direct integration to the [Web Audio API](/en-US/docs/Web/API/Web_Audio_API), however the extracted `Float32Array` samples can be copied directly into a {{domxref("AudioBuffer")}} for playback.

Reviewer 2: Cool. Do we need to mention the relationship between this and the EncodedAudioChunk too? I mean we have an image of it, but we say nothing about it.

Author: Done. I thought I had added that, I know the linter was aggressive and had at various points removed paragraphs I had previously written. I added a new paragraph here about  EncodedAudioChunk

![AudioData and EncodedAudioChunk](audio-data.png)

### Processing model

> The WebCodecs API uses an asynchronous [processing model](https://w3c.github.io/webcodecs/#codec-processing-model-section). Each instance of an encoder or decoder maintains an internal, independent processing queue. When queueing a substantial amount of work, it's important to keep this model in mind.

Reviewer 2: What is "of work"?

Reviewer 2: I mean do you mean data, or packets, or instructions, or something else. 

It matters because we didn't mention work before, and we mention it in the next few sections.

Author: I didn't actually write this section, it's already there in the live version / main branch. I did add a few words to clarify though

> Methods named `configure()`, `encode()`, `decode()`, and `flush()` operate asynchronously by appending control messages to the end of the queue, while methods named `reset()` and `close()` synchronously abort all pending work and purge the processing queue. After `reset()`, more work may be queued following a call to `configure()`, but `close()` is a permanent operation.

Reviewer 2: It would be good to link to these methods if you can? Is the reason you chose not to because they exist on both audio and video objects? If so, I'd say something like "Video methods named ..." and then at the end - "the same methods exist on the audio Xxxx and have the same behaviour".

Author: Again, I didn't write this section, this is preserved from the main branch / live version: https://developer.mozilla.org/en-US/docs/Web/API/WebCodecs_API

I did add the methods though and added a clarification at the end.

Reviewer 2: Thank you. I get reviews like this (mine) all the time on stuff I didn't touch. It's annoying, because you never read the stuff you didn't touch as carefully as the stuff you did touch.

Appreciate you fixing it up.

The `flush()` method can be used to wait for the completion of all work that was pending at the time `flush()` was called. However, it should generally only be called once all desired work is queued — it is not intended to force progress at regular intervals. Calling it unnecessarily will affect encoder quality and cause decoders to require the next input to be a key frame.

### Codecs

> A codec is a specific algorithm for encoding (compressing) and decoding (decompressing) video and audio. There are several industry standard codecs for video, and a separate set of codecs for audio. Here are the major ones supported in WebCodecs:

Reviewer 2: ```suggestion
A codec is a specific algorithm for encoding (compressing) and decoding (decompressing) video and audio. There are several industry standard codecs for video, and a separate set of codecs for audio. Here are the major ones supported by the WebCodecs API:
```

#### Video codecs

- H.264 (AVC)
  - : The most widely supported video codec. Most MP4 files use H.264.
- VP9
  - : Open source, developed by Google. Better compression than H.264. Commonly used on YouTube and in WebM files.
- AV1
  - : The newest open source codec, with better compression than VP9. Broad decoder support; hardware encoder support is still limited.
- H.265 (HEVC)
  - : Better compression than H.264, but with significant gaps in browser support outside of Apple platforms.

#### Audio codecs

- Opus
  - : Open source, low-latency. The recommended choice for most WebCodecs audio encoding.
- AAC
  - : Widely supported. Common in MP4 files.
- MP3
  - : Broadly supported for decoding, but not available as an encoder in WebCodecs.
- PCM
  - : Uncompressed audio. No quality loss, but large file sizes.

> The WebCodecs specification only supports a specific set of codecs, and individual devices and browsers may only support a subset of those. Encoders and decoders must be configured with fully specified codec strings (such as `"vp09.00.40.08.00"` for VP9 or `"avc1.4d0034"` for H.264) instead of ambiguous codec names like `"vp9"` or `"h264"`. For a full list of codec strings and their browser support, see the [Codec Support Table](https://webcodecsfundamentals.org/datasets/codec-support-table/) on WebCodecs Fundamentals, or the [Codec selection guide](/en-US/docs/Web/API/WebCodecs_API/Codec_selection) for guidance on choosing the right codec string.

Reviewer 2: The use of specification, specific, specified here in such a dense format makes it hard to read. You could perhaps do something like

```suggestion
The WebCodecs specification supports a particular set of codecs, and individual devices and browsers may only support a subset of those. Encoders and decoders must be configured with fully specified codec strings (such as `"vp09.00.40.08.00"` for VP9 or `"avc1.4d0034"` for H.264) instead of ambiguous codec names like `"vp9"` or `"h264"`. For a full list of codec strings and their browser support, see the [Codec Support Table](https://webcodecsfundamentals.org/datasets/codec-support-table/) (webcodecsfundamentals.org), or the [Codec selection guide](/en-US/docs/Web/API/WebCodecs_API/Codec_selection) for guidance on choosing the right codec string.
```

### Muxing and demuxing

The WebCodecs API only deals with encoding and decoding, with encoded chunks just representing binary data. It does not provide a built-in way to read `EncodedVideoChunk` objects from a video file, or write them to a playable video file.

Reading encoded chunks from a video file is a completely different process called demuxing, and to fetch `EncodedVideoChunk` objects from a video file, you will need to use a demuxing library such as [Mediabunny](https://mediabunny.dev/) or [web-demuxer](https://github.com/bilibili/web-demuxer).

![Demuxer](decoder-demuxer.png)

These libraries will follow the video container specifications (e.g., webm, mp4) to extract the track data and byte offsets for each encoded chunk, and provide methods for extracting the actual chunks from the raw file.

Likewise, to write to a playable video file, you will need a muxing library, with [Mediabunny](https://mediabunny.dev/) being the primary option for muxing. Muxing libraries will handle formatting the binary encoded data, and placing it in the correct byte position in the output file stream according to the container specification, so that the output video is playable.

> You can find more information on muxing and demuxing in the [Muxing and Demuxing guide](https://webcodecsfundamentals.org/basics/muxing/) on WebCodecs Fundamentals.

Reviewer 2: More common to do this:

```suggestion
You can find more information on muxing and demuxing in the [Muxing and Demuxing guide](https://webcodecsfundamentals.org/basics/muxing/) (webcodecsfundamentals.org).
```

## Guides

- [Video processing concepts](/en-US/docs/Web/API/WebCodecs_API/Video_processing_concepts)

>   - : A brief primer on video processing, including codecs and containers, muxing and demuxing, that covers conceptual information to understand how the WebCodecs API implements these concepts.

Reviewer 2: ```suggestion
  - : A brief primer on video processing, covering codecs and containers, muxing and demuxing, and provides conceptual information to understand how the WebCodecs API implements these concepts.
```

- [Using the WebCodecs API](/en-US/docs/Web/API/WebCodecs_API/Using_the_WebCodecs_API)

>   - : In depth guide to how to actually use the WebCodecs API, including how to instantiate and configure encoders and decoders, how to create and consume video frames, and how to extract samples from AudioData.

Reviewer 2: ```suggestion
  - : In depth guide to how to actually use the WebCodecs API, including how to instantiate and configure encoders and decoders, how to create and consume video frames, and how to extract samples from `AudioData`.
```

- [Codec selection](/en-US/docs/Web/API/WebCodecs_API/Codec_selection)

>   - : The WebCodecs API requires codec strings — precise identifiers that specify not just the codec family but also the profile, level, and other parameters. This guide explains how codec strings work and how to choose the right codec for common use cases.

Reviewer 2: Guides are never listed - see https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

Normally you'd just list them in the sidebar and as See also section.
IMO if we have them in-page we should have them below the interfaces.

@wbamberg Do you have any thoughts on listing guides in API Overview pages. To my mind it is an omission - sure they are guides and we separate our doc types, but it is useful to have them here in the main entry point.

Reviewer 3: No, they're not, but I agree that they should be. I think of API overview pages as like expanded versions of the sidebar really, so it makes sense to have guides listed (and I would have them first, to mirror the sidebar).

I also think we ought to be stricter about not having too much exposition in them, and tend towards moving exposition into separate guide pages. API overview pages IMO are primarily navigational, although they do also provide context and motivation. e.g. I think the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) page is OK, but the [Service Worker API](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) is too much.

Author: We could move much of the content in the index to an "Intro to WebCodecs" guide, and slim down to a much narrower concepts and usage  on the main page, just talking about the core interfaces (VideoEncoder, AudioEncoder, VideoDecoder, AudioDecoder) and the core methods (currently in processing model) and move everything else (codecs, muxing, diagrams) into an intro or concepts guide?

Reviewer 3: Yes, something like this would be my preference, but I'm not the reviewer of this, Hamish is, so only if he agrees. And as you see, MDN is not consistent about doing this, so it might just reflect my personal prejudices rather than any considered policy.

This is probably obvious, but to unpack a bit what I mean by "primarily navigational": I expect most of the time people land on this page, they are going to be looking to get somewhere else, rather than spend a lot of time reading stuff. And in a page like the [Service Worker API](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API), the exposition at the start makes it harder for people to find links to the places they might want to go, like the interface pages.

Reviewer 2: Thanks @wbamberg  (and @sb2702 )

I broadly see API overviews the same way as you. 
I'm completely happy to have the guides section, and I can add it to the corresponding template. 

The current structure guide-like material in Web API overviews is a "sensible minimum" - ensuring that every API at least has some conceptual overview. It usually contains way too much information because it is easier to slightly extend the concepts than to generate a Using document or a separate concepts doc.

My problem is that this is a more significant change in way of working. If we do this we need buy-in and a clear instruction to authors and reviewers about what they do instead. This is complicated because it might differ if you know you are getting lots of guide or if you are just doing an introduction. 
So I might say in this case I like your suggestion @sb2702, but we need a different story if you hadn't come along with these docs.

One way we might do this is modify the templates to suggest that if more than three or four paragaphs the current concepts section should be turned into a separate guide document "XXXX API Concepts". We could then prototype this for existing docs, such as Service Worker API.
We would need a very clear outline of how much depth to put in concepts vs Using and other guide materials (I've never been really certain what goes in Using vs concepts).

I'm back day after tomorrow. I'll think about this then and see how you respond. But for now, I think lets leave this as-is.

Author: Honestly, I have no preference either way. As a reader of the docs, I can see arguments for both. I'm happy to leave it as-is

Reviewer 2: Note, I will think about this. But if we do it, we can still do it as a post process.

## Examples

> The basic instantiation of a `VideoEncoder` looks like this, where you define the output callback where `EncodedVideoChunk` objects will be returned.

Reviewer 2: The template for web APIs now requires a level 3 heading. Something like

```suggestion
### Basic usage

To  instantiate a `VideoEncoder` we pass an object that specifies a callback function that will be called when `EncodedVideoChunk` instances are available for processing, and an error function that will be called if there are errors.
This is shown in the following code:
```

```js
const encoder = new VideoEncoder({
  output(chunk, meta) {
    // Do something with chunk, typically send to muxing library
  },
  error(e) {
    console.warn(e);
  },
});
```

You then need to configure the encoder with the codec parameter and various other fields.

```js
encoder.configure({
  codec: "vp09.00.40.08.00", // See codec selection guide
  width: 1280,
  height: 720,
  bitrate: 1_000_000, // 1 Mbps
  framerate: 25,
});
```

You can then start encoding frames to the encoder. You can construct a `VideoFrame` from a `Canvas`

```js
for (let i = 0; i < 60; i++) {

>   const frame = new VideoFrame(canvas, { timestamp: (i * 1e6) / 30 }); //30 fps, in microseconds

Reviewer 4: above the encoder was configured for 25fps, not 30

Author: I adjusted the encoder example to 30fps
